File: backend/services/kiteService.js
const KiteConnect = require('kiteconnect').KiteConnect;

class KiteService {
  constructor() {
    this.apiKey = process.env.KITE_API_KEY;
    this.apiSecret = process.env.KITE_API_SECRET;
    
    if (!this.apiKey || !this.apiSecret) {
      throw new Error('Kite API credentials not found in environment variables');
    }
    
    // Initialize Kite Connect instance
    this.kc = new KiteConnect({
      api_key: this.apiKey,
      debug: process.env.NODE_ENV === 'development'
    });
    
    console.log('KiteService initialized with API Key:', this.apiKey);
    console.log('Redirect URL is configured in Kite Connect dashboard (not controlled by backend)');
  }

  // Generate login URL for OAuth flow
  getLoginUrl() {
    try {
      // Use KiteConnect's official method to generate login URL
      // The redirect URL is configured in Kite Connect dashboard, not here
      const loginUrl = this.kc.getLoginURL();
      console.log('Generated Kite login URL:', loginUrl);
      console.log('Redirect URL is configured in Kite Connect dashboard');
      
      return loginUrl;
    } catch (error) {
      console.error('Error generating Kite login URL:', error);
      throw error;
    }
  }

  // Generate session from request token
  async generateSession(requestToken) {
    try {
      console.log('Generating Kite session with request token:', requestToken);
      console.log('Using API Key:', this.apiKey);
      console.log('API Secret configured:', !!this.apiSecret);
      
      const response = await this.kc.generateSession(requestToken, this.apiSecret);
      console.log('Kite session generated successfully:', response);
      
      // Set access token for subsequent API calls
      this.kc.setAccessToken(response.access_token);
      
      return {
        accessToken: response.access_token,
        publicToken: response.public_token,
        userId: response.user_id,
        userShortname: response.user_shortname,
        userName: response.user_name,
        userType: response.user_type,
        email: response.email,
        broker: response.broker,
        exchanges: response.exchanges,
        products: response.products,
        orderTypes: response.order_types,
        loginTime: response.login_time
      };
    } catch (error) {
      console.error('Error generating Kite session:', error);
      
      // Enhanced error logging for debugging
      if (error.response && error.response.data) {
        console.error('Kite API Error Response:', error.response.data);
      }
      
      // Provide more specific error messages
      if (error.message && error.message.includes('user is not enabled')) {
        const enhancedError = new Error(`Kite Connect Error: The user is not enabled for this app. 
        
This usually means:
1. Your Kite Connect app is in development mode and this user is not in the test users list
2. Add the user's Zerodha User ID to the 'Test Users' list in your Kite Connect app settings
3. Or request production approval from Zerodha to allow all users

API Key: ${this.apiKey}
Original Error: ${error.message}`);
        
        enhancedError.originalError = error;
        throw enhancedError;
      }
      
      throw error;
    }
  }

  // Set access token for API calls
  setAccessToken(accessToken) {
    this.kc.setAccessToken(accessToken);
  }

  // Fetch user profile
  async getUserProfile() {
    try {
      const profile = await this.kc.getProfile();
      console.log('Fetched Kite user profile:', profile);
      return profile;
    } catch (error) {
      console.error('Error fetching Kite user profile:', error);
      throw error;
    }
  }

  // Fetch portfolio holdings
  async getPortfolioHoldings() {
    try {
      console.log('Fetching Kite portfolio holdings...');
      const holdings = await this.kc.getHoldings();
      console.log('Fetched Kite holdings:', holdings.length, 'positions');
      return holdings;
    } catch (error) {
      console.error('Error fetching Kite portfolio holdings:', error);
      throw error;
    }
  }

  // Fetch mutual fund holdings
  async getMutualFundHoldings() {
    try {
      console.log('Fetching Kite mutual fund holdings...');
      const mfHoldings = await this.kc.getMFHoldings();
      console.log('Fetched Kite MF holdings:', mfHoldings.length, 'funds');
      return mfHoldings;
    } catch (error) {
      console.error('Error fetching Kite mutual fund holdings:', error);
      throw error;
    }
  }

  // Get margins
  async getMargins() {
    try {
      const margins = await this.kc.getMargins();
      console.log('Fetched Kite margins:', margins);
      return margins;
    } catch (error) {
      console.error('Error fetching Kite margins:', error);
      throw error;
    }
  }

  // Transform Kite holdings to our portfolio format
  transformKiteToPortfolio(kiteHoldings, kiteMFHoldings, userProfile) {
    try {
      console.log('Transforming Kite data to portfolio format...');
      
      // Transform stock holdings
      const stocks = kiteHoldings.map(holding => {
        const currentValue = holding.quantity * holding.last_price;
        const investmentAmount = holding.quantity * holding.average_price;
        const gainLoss = currentValue - investmentAmount;
        const gainLossPercentage = investmentAmount > 0 ? (gainLoss / investmentAmount) * 100 : 0;

        return {
          symbol: holding.tradingsymbol,
          company_name: holding.tradingsymbol, // Kite doesn't provide full company name
          quantity: holding.quantity,
          avg_purchase_price: holding.average_price,
          current_price: holding.last_price,
          investment_amount: Math.round(investmentAmount),
          current_value: Math.round(currentValue),
          gain_loss: Math.round(gainLoss),
          gain_loss_percentage: Math.round(gainLossPercentage * 100) / 100,
          sector: this.getSectorFromSymbol(holding.tradingsymbol),
          exchange: holding.exchange,
          product: holding.product,
          instrument_token: holding.instrument_token,
          isin: holding.isin
        };
      });

      // Transform mutual fund holdings
      const mutualFunds = kiteMFHoldings.map(mf => {
        const currentValue = mf.quantity * mf.last_price;
        const investmentAmount = mf.quantity * mf.average_price;
        const gainLoss = currentValue - investmentAmount;
        const gainLossPercentage = investmentAmount > 0 ? (gainLoss / investmentAmount) * 100 : 0;

        return {
          scheme_name: mf.fund || mf.tradingsymbol, // Use 'fund' field for proper scheme name
          scheme_code: mf.tradingsymbol, // Keep tradingsymbol as scheme code
          units: mf.quantity,
          nav: mf.last_price,
          current_value: Math.round(currentValue),
          investment_amount: Math.round(investmentAmount),
          gain_loss: Math.round(gainLoss),
          gain_loss_percentage: Math.round(gainLossPercentage * 100) / 100,
          category: this.getMFCategoryFromSymbol(mf.fund || mf.tradingsymbol),
          fund_house: this.getFundHouseFromSymbol(mf.fund || mf.tradingsymbol),
          isin: mf.isin,
          folio_number: mf.folio || `FOLIO${Date.now()}`
        };
      });

      // Calculate portfolio summary
      const stockInvestment = stocks.reduce((sum, stock) => sum + stock.investment_amount, 0);
      const stockCurrentValue = stocks.reduce((sum, stock) => sum + stock.current_value, 0);
      const mfInvestment = mutualFunds.reduce((sum, mf) => sum + mf.investment_amount, 0);
      const mfCurrentValue = mutualFunds.reduce((sum, mf) => sum + mf.current_value, 0);
      
      const totalInvestment = stockInvestment + mfInvestment;
      const totalCurrentValue = stockCurrentValue + mfCurrentValue;
      const totalGainLoss = totalCurrentValue - totalInvestment;
      const gainLossPercentage = totalInvestment > 0 ? (totalGainLoss / totalInvestment) * 100 : 0;

      // Calculate asset allocation
      const stocksPercentage = totalCurrentValue > 0 ? (stockCurrentValue / totalCurrentValue) * 100 : 0;
      const mutualFundsPercentage = totalCurrentValue > 0 ? (mfCurrentValue / totalCurrentValue) * 100 : 0;

      const portfolioSummary = {
        total_investment: totalInvestment,
        total_current_value: totalCurrentValue,
        total_gain_loss: totalGainLoss,
        gain_loss_percentage: Math.round(gainLossPercentage * 100) / 100,
        updated_at: new Date().toISOString(),
        asset_allocation: {
          stocks: {
            value: stockCurrentValue,
            percentage: Math.round(stocksPercentage * 100) / 100
          },
          mutual_funds: {
            value: mfCurrentValue,
            percentage: Math.round(mutualFundsPercentage * 100) / 100
          }
        }
      };

      console.log('Portfolio transformation completed. Summary:', portfolioSummary);

      return {
        stocks,
        mutual_funds: mutualFunds,
        summary: portfolioSummary
      };
    } catch (error) {
      console.error('Error transforming Kite data to portfolio format:', error);
      throw error;
    }
  }

  // Helper function to determine sector from symbol (basic mapping)
  getSectorFromSymbol(symbol) {
    const sectorMapping = {
      'TCS': 'Information Technology',
      'INFY': 'Information Technology',
      'WIPRO': 'Information Technology',
      'HCLTECH': 'Information Technology',
      'TECHM': 'Information Technology',
      'HDFCBANK': 'Banking',
      'ICICIBANK': 'Banking',
      'SBIN': 'Banking',
      'AXISBANK': 'Banking',
      'KOTAKBANK': 'Banking',
      'RELIANCE': 'Oil & Gas',
      'ONGC': 'Oil & Gas',
      'BPCL': 'Oil & Gas',
      'IOC': 'Oil & Gas',
      'HINDUNILVR': 'FMCG',
      'ITC': 'FMCG',
      'NESTLEIND': 'FMCG',
      'BRITANNIA': 'FMCG',
      'BAJFINANCE': 'Financial Services',
      'BAJAJFINSV': 'Financial Services',
      'LT': 'Construction',
      'ULTRACEMCO': 'Cement',
      'ASIANPAINT': 'Paints',
      'MARUTI': 'Automobile',
      'M&M': 'Automobile',
      'TATAMOTORS': 'Automobile'
    };

    // Extract base symbol (remove suffixes like -EQ, -BE etc.)
    const baseSymbol = symbol.split('-')[0];
    return sectorMapping[baseSymbol] || 'Other';
  }

  // Helper function to determine MF category from scheme name
  getMFCategoryFromSymbol(schemeName) {
    if (!schemeName) return 'Other';
    
    const lowerName = schemeName.toLowerCase();
    
    // Order matters - check more specific categories first
    if (lowerName.includes('elss') || lowerName.includes('tax saver')) return 'ELSS';
    if (lowerName.includes('liquid')) return 'Liquid Fund';
    if (lowerName.includes('ultra short') || lowerName.includes('money market')) return 'Ultra Short Duration Fund';
    if (lowerName.includes('short') && lowerName.includes('duration')) return 'Short Duration Fund';
    if (lowerName.includes('medium') && lowerName.includes('duration')) return 'Medium Duration Fund';
    if (lowerName.includes('long') && lowerName.includes('duration')) return 'Long Duration Fund';
    if (lowerName.includes('gilt') || lowerName.includes('government')) return 'Gilt Fund';
    if (lowerName.includes('corporate bond') || lowerName.includes('credit')) return 'Corporate Bond Fund';
    if (lowerName.includes('debt') || lowerName.includes('bond')) return 'Debt Fund';
    if (lowerName.includes('balanced') || lowerName.includes('hybrid')) return 'Hybrid Fund';
    if (lowerName.includes('aggressive hybrid')) return 'Aggressive Hybrid Fund';
    if (lowerName.includes('conservative hybrid')) return 'Conservative Hybrid Fund';
    if (lowerName.includes('large cap') || lowerName.includes('bluechip')) return 'Large Cap Fund';
    if (lowerName.includes('mid cap')) return 'Mid Cap Fund';
    if (lowerName.includes('small cap')) return 'Small Cap Fund';
    if (lowerName.includes('multi cap') || lowerName.includes('multicap')) return 'Multi Cap Fund';
    if (lowerName.includes('flexi cap') || lowerName.includes('flexicap')) return 'Flexi Cap Fund';
    if (lowerName.includes('sectoral') || lowerName.includes('thematic')) return 'Sectoral/Thematic Fund';
    if (lowerName.includes('index') || lowerName.includes('nifty') || lowerName.includes('sensex')) return 'Index Fund';
    if (lowerName.includes('equity')) return 'Equity Fund';
    if (lowerName.includes('international') || lowerName.includes('global')) return 'International Fund';
    
    return 'Other';
  }

  // Helper function to extract fund house from symbol/name
  getFundHouseFromSymbol(symbol) {
    if (!symbol) return 'Unknown Fund House';
    
    const upperSymbol = symbol.toUpperCase();
    const fundHouseMapping = {
      'SBI': 'SBI Mutual Fund',
      'HDFC': 'HDFC Mutual Fund',
      'ICICI': 'ICICI Prudential Mutual Fund',
      'AXIS': 'Axis Mutual Fund',
      'KOTAK': 'Kotak Mutual Fund',
      'NIPPON': 'Nippon India Mutual Fund',
      'ADITYA BIRLA': 'Aditya Birla Sun Life Mutual Fund',
      'BIRLA SUN LIFE': 'Aditya Birla Sun Life Mutual Fund',
      'UTI': 'UTI Mutual Fund',
      'DSP': 'DSP Mutual Fund',
      'FRANKLIN': 'Franklin Templeton Mutual Fund',
      'INVESCO': 'Invesco Mutual Fund',
      'MIRAE': 'Mirae Asset Mutual Fund',
      'TATA': 'Tata Mutual Fund',
      'RELIANCE': 'Reliance Mutual Fund',
      'MAHINDRA': 'Mahindra Mutual Fund',
      'L&T': 'L&T Mutual Fund',
      'MOTILAL': 'Motilal Oswal Mutual Fund',
      'EDELWEISS': 'Edelweiss Mutual Fund',
      'CANARA': 'Canara Robeco Mutual Fund',
      'ROBECO': 'Canara Robeco Mutual Fund',
      'BARODA': 'Baroda BNP Paribas Mutual Fund',
      'BNP PARIBAS': 'Baroda BNP Paribas Mutual Fund'
    };

    for (const [key, value] of Object.entries(fundHouseMapping)) {
      if (upperSymbol.includes(key)) {
        return value;
      }
    }
    return 'Unknown Fund House';
  }

  // Create user profile from Kite data
  createUserProfile(kiteProfile, sessionData) {
    return {
      user_profile: {
        name: kiteProfile.user_name || sessionData.userName,
        email: kiteProfile.email || sessionData.email,
        age: 30, // Default age as Kite doesn't provide this
        profession: 'Investor',
        location: 'India',
        kite_user_id: kiteProfile.user_id || sessionData.userId
      },
      financial_profile: {
        monthly_salary: 100000, // Default values as Kite doesn't provide this
        annual_ctc: 1200000,
        take_home: 85000
      },
      investment_profile: {
        risk_tolerance: 'moderate', // Default, can be updated based on portfolio analysis
        investment_experience: 'intermediate',
        investment_horizon: '5-10 years'
      }
    };
  }
}

module.exports = KiteService;

File: backend_python/services/kite_service.py
"""
Kite Connect integration service for live portfolio data
"""

import logging
from typing import Dict, Any, List, Optional, Tuple
from datetime import datetime
from kiteconnect import KiteConnect

from config.settings import get_settings
from models.user import UserProfile, FinancialProfile, InvestmentProfile, Portfolio, StockHolding, MutualFundHolding
from models.portfolio import KiteHolding, KiteMFHolding

logger = logging.getLogger(__name__)
settings = get_settings()


class KiteService:
    def __init__(self):
        if not settings.kite_api_key or not settings.kite_api_secret:
            raise ValueError("Kite API credentials not found in environment variables")
        
        self.api_key = settings.kite_api_key
        self.api_secret = settings.kite_api_secret
        
        # Initialize Kite Connect instance
        self.kite = KiteConnect(api_key=self.api_key)
        
        logger.info(f"KiteService initialized with API Key: {self.api_key}")
        logger.info("Redirect URL is configured in Kite Connect dashboard")
    
    def get_login_url(self) -> str:
        """Generate login URL for OAuth flow"""
        try:
            login_url = self.kite.login_url()
            logger.info(f"Generated Kite login URL: {login_url}")
            return login_url
        except Exception as e:
            logger.error(f"Error generating Kite login URL: {str(e)}")
            raise
    
    async def generate_session(self, request_token: str) -> Dict[str, Any]:
        """Generate session from request token"""
        try:
            logger.info(f"Generating Kite session with request token: {request_token}")
            
            # Generate session
            data = self.kite.generate_session(request_token, api_secret=self.api_secret)
            
            logger.info("Kite session generated successfully")
            
            return {
                "accessToken": data["access_token"],
                "publicToken": data.get("public_token"),
                "userId": data["user_id"],
                "userShortname": data.get("user_shortname"),
                "email": data.get("email"),
                "loginTime": datetime.now().isoformat()
            }
        
        except Exception as e:
            logger.error(f"Error generating Kite session: {str(e)}")
            raise
    
    def set_access_token(self, access_token: str):
        """Set access token for API calls"""
        self.kite.set_access_token(access_token)
    
    async def get_user_profile(self) -> Dict[str, Any]:
        """Get user profile from Kite"""
        try:
            profile = self.kite.profile()
            logger.info(f"Fetched Kite user profile for user: {profile.get('user_id')}")
            return profile
        except Exception as e:
            logger.error(f"Error fetching Kite user profile: {str(e)}")
            raise
    
    async def get_portfolio_holdings(self) -> List[Dict[str, Any]]:
        """Get stock portfolio holdings from Kite"""
        try:
            holdings = self.kite.holdings()
            logger.info(f"Fetched {len(holdings)} stock holdings from Kite")
            return holdings
        except Exception as e:
            logger.error(f"Error fetching Kite portfolio holdings: {str(e)}")
            raise
    
    async def get_mutual_fund_holdings(self) -> List[Dict[str, Any]]:
        """Get mutual fund holdings from Kite"""
        try:
            # Note: Kite Connect may have limited MF support, this is a placeholder
            # In practice, you might need to use different endpoints or services
            try:
                mf_holdings = self.kite.holdings()  # Filter for MF if available
                # Filter out non-MF holdings if needed
                mf_only = [h for h in mf_holdings if 'MF' in h.get('exchange', '')]
                logger.info(f"Fetched {len(mf_only)} mutual fund holdings from Kite")
                return mf_only
            except:
                # Return empty list if MF holdings not available
                logger.warning("Mutual fund holdings not available through Kite Connect")
                return []
        except Exception as e:
            logger.error(f"Error fetching Kite MF holdings: {str(e)}")
            return []
    
    def transform_kite_to_portfolio(self, holdings: List[Dict[str, Any]], 
                                  mf_holdings: List[Dict[str, Any]], 
                                  kite_profile: Dict[str, Any]) -> Portfolio:
        """Transform Kite holdings to our portfolio format"""
        
        # Convert stock holdings
        stocks = []
        total_stock_investment = 0
        total_stock_value = 0
        
        for holding in holdings:
            try:
                # Skip if it's actually a mutual fund
                if 'MF' in holding.get('exchange', ''):
                    continue
                
                quantity = holding.get('quantity', 0)
                if quantity <= 0:
                    continue
                
                avg_price = holding.get('average_price', 0)
                current_price = holding.get('last_price', avg_price)
                
                investment_amount = quantity * avg_price
                current_value = quantity * current_price
                gain_loss = current_value - investment_amount
                gain_loss_percentage = (gain_loss / investment_amount * 100) if investment_amount > 0 else 0
                
                stock = {
                    "symbol": holding.get('tradingsymbol', ''),
                    "company_name": holding.get('tradingsymbol', '').replace('-EQ', ''),
                    "quantity": quantity,
                    "avg_cost": avg_price,
                    "current_price": current_price,
                    "investment_amount": investment_amount,
                    "current_value": current_value,
                    "gain_loss": gain_loss,
                    "gain_loss_percentage": gain_loss_percentage,
                    "sector": self._get_sector_from_symbol(holding.get('tradingsymbol', ''))
                }
                
                stocks.append(stock)
                total_stock_investment += investment_amount
                total_stock_value += current_value
                
            except Exception as e:
                logger.warning(f"Error processing stock holding: {str(e)}")
                continue
        
        # Convert mutual fund holdings
        mutual_funds = []
        total_mf_investment = 0
        total_mf_value = 0
        
        for mf_holding in mf_holdings:
            try:
                quantity = mf_holding.get('quantity', 0)
                if quantity <= 0:
                    continue
                
                avg_price = mf_holding.get('average_price', 0)
                current_nav = mf_holding.get('last_price', avg_price)
                
                investment_amount = quantity * avg_price
                current_value = quantity * current_nav
                gain_loss = current_value - investment_amount
                gain_loss_percentage = (gain_loss / investment_amount * 100) if investment_amount > 0 else 0
                
                mf = {
                    "scheme_name": mf_holding.get('tradingsymbol', ''),
                    "folio_number": mf_holding.get('folio', ''),
                    "units": quantity,
                    "nav": current_nav,
                    "investment_amount": investment_amount,
                    "current_value": current_value,
                    "gain_loss": gain_loss,
                    "gain_loss_percentage": gain_loss_percentage,
                    "sip_amount": 0,  # Not available from Kite
                    "fund_type": "Equity",  # Default
                    "fund_category": "Mixed"
                }
                
                mutual_funds.append(mf)
                total_mf_investment += investment_amount
                total_mf_value += current_value
                
            except Exception as e:
                logger.warning(f"Error processing MF holding: {str(e)}")
                continue
        
        # Calculate totals
        total_investment = total_stock_investment + total_mf_investment
        total_current_value = total_stock_value + total_mf_value
        total_gain_loss = total_current_value - total_investment
        gain_loss_percentage = (total_gain_loss / total_investment * 100) if total_investment > 0 else 0
        
        # Asset allocation
        asset_allocation = {}
        
        if total_current_value > 0:
            if total_stock_value > 0:
                asset_allocation["stocks"] = {
                    "value": total_stock_value,
                    "percentage": (total_stock_value / total_current_value * 100)
                }
            
            if total_mf_value > 0:
                asset_allocation["mutual_funds"] = {
                    "value": total_mf_value,
                    "percentage": (total_mf_value / total_current_value * 100)
                }
        
        # Create portfolio summary
        portfolio_summary = {
            "total_investment": total_investment,
            "total_current_value": total_current_value,
            "total_gain_loss": total_gain_loss,
            "gain_loss_percentage": gain_loss_percentage,
            "asset_allocation": asset_allocation,
            "updated_at": datetime.now().isoformat()
        }
        
        return {
            "summary": portfolio_summary,
            "stocks": stocks,
            "mutual_funds": mutual_funds
        }
    
    def create_user_profile(self, kite_profile: Dict[str, Any], session_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create user profile from Kite profile data"""
        
        # Extract user information from Kite profile
        user_name = kite_profile.get('user_name', session_data.get('userShortname', 'Kite User'))
        email = kite_profile.get('email', session_data.get('email', ''))
        
        # Estimate financial profile based on portfolio (rough estimates)
        estimated_income = 1000000  # Default 10L annual income
        
        user_profile = {
            "user_profile": {
                "name": user_name,
                "age": 35,  # Default age
                "profession": "Investor",
                "location": "India",
                "email": email,
                "phone": ""
            },
            "financial_profile": {
                "annual_income_after_tax": estimated_income,
                "monthly_income_after_tax": estimated_income / 12,
                "take_home": estimated_income / 12,
                "monthly_expenses": estimated_income * 0.6 / 12,
                "savings_rate": 0.4
            },
            "investment_profile": {
                "risk_tolerance": "Moderate",
                "investment_experience": "Experienced",  # Assuming Kite users are experienced
                "investment_horizon": 10,
                "preferred_investment_types": ["Stocks", "Mutual Funds"]
            }
        }
        
        return user_profile
    
    def _get_sector_from_symbol(self, symbol: str) -> str:
        """Get sector from stock symbol (simplified mapping)"""
        sector_mapping = {
            'INFY': 'Information Technology',
            'TCS': 'Information Technology',
            'HDFCBANK': 'Banking',
            'ICICIBANK': 'Banking',
            'RELIANCE': 'Oil & Gas',
            'WIPRO': 'Information Technology',
            'ITC': 'FMCG',
            'HINDUNILVR': 'FMCG',
            'BHARTIARTL': 'Telecommunications',
            'SBIN': 'Banking'
        }
        
        # Remove exchange suffix if present
        clean_symbol = symbol.replace('-EQ', '').replace('-BE', '')
        
        return sector_mapping.get(clean_symbol, 'Other')
    
    async def refresh_portfolio_data(self, user_id: str) -> Dict[str, Any]:
        """Refresh portfolio data for a Kite user"""
        try:
            # Fetch fresh data
            holdings = await self.get_portfolio_holdings()
            mf_holdings = await self.get_mutual_fund_holdings()
            
            # Get user profile (cached)
            user_profile = await self.get_user_profile()
            
            # Transform to portfolio format
            portfolio = self.transform_kite_to_portfolio(holdings, mf_holdings, user_profile)
            
            logger.info(f"Refreshed portfolio data for Kite user: {user_id}")
            
            return {
                "success": True,
                "portfolio": portfolio,
                "updated_at": datetime.now().isoformat()
            }
        
        except Exception as e:
            logger.error(f"Error refreshing portfolio data: {str(e)}")
            return {
                "success": False,
                "error": str(e)
            }